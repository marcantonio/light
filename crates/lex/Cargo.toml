[package]
name = "lex"
version = "0.0.0"
edition = "2021"

[dependencies]
itertools = "0.10.5"
serde = { version = "1.0.136", features = ["derive"] }

common = { path = "../common" }

[dev-dependencies]
insta = { version = "1.12.0", features = ["glob"] }
