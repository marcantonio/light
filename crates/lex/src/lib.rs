use serde::Serialize;
use std::iter::Peekable;

use common::Operator;
pub use token::{Token, TokenType};

#[cfg(test)]
mod tests;
pub mod token;

pub type LexResult = std::result::Result<Token, LexError>;

pub struct Lex {
    stream: Peekable<StreamIter<char>>,
    pub tokens: Vec<Token>,
}

impl Lex {
    pub fn new(input: &str) -> Self {
        Lex { stream: StreamIter::new(input).peekable(), tokens: vec![] }
    }

    // Scan all input
    pub fn scan(mut self) -> Result<Vec<Token>, LexError> {
        loop {
            let token = self.lex()?;
            if token.is_eof() {
                break;
            }
            self.tokens.push(token);
        }
        Ok(self.tokens)
    }

    // Recursively process enough characters to produce one token
    fn lex(&mut self) -> LexResult {
        use TokenType::*;

        let cur = match self.stream.next() {
            Some(cur) => cur,
            None => unreachable!("can't lex nothing"),
        };

        // Inject a semicolon if certain tokens occur at the end of the line or
        // EOF. If EOF, make sure the context is right.
        if cur.value == '\n' && self.should_add_semicolon() {
            return Ok(Token::new(Semicolon(true), cur.line, cur.column));
        } else if cur.is_eof() {
            if self.should_add_semicolon() {
                let semi = match self.tokens.last() {
                    Some(t) => Token::new(Semicolon(true), t.line, t.column + 1),
                    None => Token::default(),
                };
                self.tokens.push(semi);
            }
            return Ok(Token::new(Eof, cur.line, cur.column));
        }

        // Ignore whitespace
        if cur.value.is_ascii_whitespace() {
            while let Some(c) = self.stream.peek() {
                if !c.value.is_ascii_whitespace() {
                    return self.lex();
                } else if c.is_eof() {
                    break;
                }
                self.stream.next();
            }
            return self.lex(); // Eat trailing newline
        }

        // Single line comments
        if cur == '/' && matches!(self.stream.peek(), Some(c) if *c == '/') {
            while let Some(c) = self.stream.next() {
                if c == '\n' {
                    return self.lex();
                } else if c.is_eof() {
                    break;
                }
            }
            return self.lex(); // Eat trailing comment
        }

        // Keywords, types, and identifiers
        if cur.value.is_ascii_alphabetic() {
            let mut identifier = String::from(cur.value);
            while let Some(c) = self.stream.peek() {
                if c.value.is_ascii_alphanumeric() || *c == '_' {
                    identifier.push(c.value);
                    self.stream.next();
                } else {
                    break;
                }
            }

            let tt = match identifier.as_str() {
                "fn" => Fn,
                "let" => Let,
                "for" => For,
                "if" => If,
                "else" => Else,
                "extern" => Extern,
                "true" => Bool(true),
                "false" => Bool(false),
                "struct" => Struct,
                "module" => Module,
                _ => Ident(identifier),
            };

            return Ok(Token::new(tt, cur.line, cur.column));
        }

        // Literal numbers
        if cur.value.is_ascii_digit() {
            let mut n = String::from(cur.value);
            while let Some(c) = self.stream.peek() {
                if c.value.is_ascii_alphanumeric() || *c == '.' {
                    n.push(c.value);
                    self.stream.next();
                } else {
                    break;
                }
            }

            return Ok(Token::new(Num(n), cur.line, cur.column));
        }

        // Literal char
        if cur == '\'' {
            let mut ch = String::new();
            let next = self.stream.next().unwrap_or_else(|| unreachable!("lexed None when looking for char"));

            match next.value {
                // Control characters
                '\\' => {
                    if let Some(next) = self.stream.next() {
                        match next.value {
                            'n' => ch = String::from("\n"),
                            't' => ch = String::from("\t"),
                            '\'' => ch = String::from("'"),
                            c => {
                                return Err(LexError::from((
                                    format!("Invalid character control sequence: `\\{}`", c),
                                    next,
                                )))
                            },
                        }
                    }
                },
                // EOF
                '\0' => {
                    return Err(LexError::from((
                        "Unterminated character literal. Expecting `'`, got `EOF`".to_string(),
                        cur,
                    )));
                },
                '\'' => return Err(LexError::from(("Character literal can't be empty".to_string(), cur))),

                // Everything else
                c => ch = String::from(c),
            }

            // Check for closing '\''
            let last =
                self.stream.next().unwrap_or_else(|| unreachable!("fatal: lexed None when looking for `'`"));
            match last.value {
                '\'' => (),
                '\0' | '\n' => {
                    return Err(LexError::from((
                        "Unterminated character literal. Expecting `'`".to_string(),
                        last,
                    )));
                },
                _ => {
                    return Err(LexError::from((
                        format!("Invalid character sequence: `'{}{}'`", ch, last.value),
                        last,
                    )));
                },
            }

            return Ok(Token::new(Char(ch), cur.line, cur.column));
        }

        // Multi-character operators
        if let Some(next) = self.stream.peek() {
            match cur.value {
                '=' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::Eq), cur.line, cur.column));
                },
                '!' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::NotEq), cur.line, cur.column));
                },
                '>' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::GtEq), cur.line, cur.column));
                },
                '<' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::LtEq), cur.line, cur.column));
                },
                '&' if next == &'&' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::And), cur.line, cur.column));
                },
                '|' if next == &'|' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::Or), cur.line, cur.column));
                },
                '+' if next == &'+' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::Inc), cur.line, cur.column));
                },
                '+' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::AddEq), cur.line, cur.column));
                },
                '-' if next == &'-' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::Dec), cur.line, cur.column));
                },
                '-' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::SubEq), cur.line, cur.column));
                },
                '-' if next == &'>' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::RetType), cur.line, cur.column));
                },
                '*' if next == &'*' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::Pow), cur.line, cur.column));
                },
                '*' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::MulEq), cur.line, cur.column));
                },
                '/' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::DivEq), cur.line, cur.column));
                },
                _ => (),
            }
        }

        // Everything else
        let tt = match cur.value {
            '+' => Op(Operator::Add),
            '-' => Op(Operator::Sub),
            '*' => Op(Operator::Mul),
            '/' => Op(Operator::Div),
            '>' => Op(Operator::Gt),
            '<' => Op(Operator::Lt),
            '!' => Op(Operator::Not),
            '=' => Op(Operator::Assign),
            '&' => Op(Operator::BitAnd),
            '^' => Op(Operator::BitXor),
            '|' => Op(Operator::BitOr),
            '.' => Dot,
            '}' => CloseBrace,
            ']' => CloseBracket,
            ')' => CloseParen,
            ':' => Colon,
            ',' => Comma,
            '{' => OpenBrace,
            '[' => OpenBracket,
            '(' => OpenParen,
            ';' => Semicolon(false),
            c => {
                return Err(LexError::from((format!("Unknown character: {}", c), cur)));
            },
        };

        Ok(Token::new(tt, cur.line, cur.column))
    }

    // Add a semicolon for these tokens
    fn should_add_semicolon(&self) -> bool {
        use TokenType::*;

        if let Some(t) = self.tokens.last() {
            matches!(
                t.tt,
                Bool(_)
                    | Char(_)
                    | CloseBrace
                    | CloseParen
                    | CloseBracket
                    | Ident(_)
                    | Num(_)
                    | Op(Operator::Inc)
                    | Op(Operator::Dec)
            )
        } else {
            false
        }
    }
}

// Provides additional context for each source character
#[derive(Debug, Clone, Copy, PartialEq)]
struct ContextElement<T> {
    value: T,
    line: usize,
    column: usize,
}

impl<T> ContextElement<T> {
    fn new(value: T, line: usize, column: usize) -> Self {
        ContextElement { value, line: line + 1, column: column + 1 }
    }
}

impl ContextElement<char> {
    // Will cause lexing to stop if there's a null byte in the file
    fn is_eof(&self) -> bool {
        self.value == 0 as char
    }
}

impl PartialEq<char> for ContextElement<char> {
    fn eq(&self, other: &char) -> bool {
        self.value == *other
    }
}

// Iterator for the source character stream. Supports line and column context.
struct StreamIter<T> {
    lines: Vec<Vec<T>>,
    line: usize,
    column: usize,
}

impl StreamIter<char> {
    fn new(input: &str) -> Self {
        StreamIter {
            lines: input
                .split_inclusive('\n') // can't use .lines() or we lose '\n'
                .map(|l| l.chars().collect())
                .collect(),
            line: 0,
            column: 0,
        }
    }
}

impl Iterator for StreamIter<char> {
    type Item = ContextElement<char>;

    fn next(&mut self) -> Option<Self::Item> {
        let opt = self.lines.get(self.line);
        let line = match opt {
            Some(l) => l,
            None => return Some(ContextElement::new(0 as char, self.line, self.column - 1)),
        };
        let cc =
            line.get(self.column).map(|c| ContextElement::new(*c, self.line, self.column)).or_else(|| {
                self.line += 1;
                self.column = 0;
                self.lines.get(self.line).and_then(|line| {
                    line.get(self.column).map(|c| ContextElement::new(*c, self.line, self.column))
                })
            });
        self.column += 1;
        cc.or_else(|| Some(ContextElement::new(0 as char, self.line, self.column - 1)))
    }
}

#[derive(Debug, PartialEq, Eq, Serialize)]
pub struct LexError {
    message: String,
    line: usize,
    column: usize,
}

impl std::fmt::Display for LexError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Lexing error: {} at {}:{}", self.message, self.line, self.column)
    }
}

impl std::error::Error for LexError {}

impl<T> From<(String, ContextElement<T>)> for LexError {
    fn from((msg, cp): (String, ContextElement<T>)) -> Self {
        LexError { message: msg, line: cp.line, column: cp.column }
    }
}
