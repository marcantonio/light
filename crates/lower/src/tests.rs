use super::*;
use lex::Lex;
use parse::Parse;
use tych::Tych;

macro_rules! run_insta {
    ($prefix:expr, $tests:expr) => {
        insta::with_settings!({ snapshot_path => "tests/snapshots", prepend_module_to_snapshot => false }, {
            for test in $tests {
                let tokens = Lex::new(test[1]).scan().expect("lexing faled in `lower` tests");
                let (ast, mut symbol_table, _, _) = Parse::new(&tokens).parse().expect("parsing failed in `lower` tests");
                let typed_ast = Tych::new(&mut symbol_table).walk(ast).expect("type checking failed in `lower` tests");
                let res = Lower::new(vec![], &mut symbol_table).walk(typed_ast);
                insta::assert_yaml_snapshot!(format!("{}_{}", $prefix, test[0]), (test[1], res));
            }
        })
    };
}

#[test]
fn test_binop() {
    let tests = [
        [
            "all_ops",
            r#"
fn main() {
    let y: int
    y += 1
    y -= 1
    y *= 1
    y /= 1
    y <= 1
}
"#,
        ],
        [
            "inner",
            r#"
fn main() {
    let x: int = 1
    let y: bool = x <= 3
}
"#,
        ],
    ];
    run_insta!("binop", tests);
}

#[test]
fn test_func_call() {
    let tests = [
        [
            "def_first",
            r#"
fn foo() {}

fn main() {
    foo()
}
"#,
        ],
        [
            "def_second",
            r#"
fn main() {
    foo()
}

fn foo() {}
"#,
        ],
        [
            "mix1",
            r#"
fn main() {
    foo()
}

fn foo() {}

fn bar() {
    foo()
}
"#,
        ],
        [
            "mix2",
            r#"
fn main() {
    foo()
    foo()
}

fn foo() {}

fn bar() {
    foo()
}
"#,
        ],
        [
            "mix3",
            r#"
fn main() {
    foo()
}

fn foo() {}

fn bar() {
    foo()
    foo()
}
"#,
        ],
        [
            "recursive",
            r#"
fn foo() {
    foo()
}

fn main() {
    foo()
}
"#,
        ],
    ];
    run_insta!("func_call", tests);
}

#[test]
fn test_struct() {
    let tests = [
        [
            "selector",
            r#"
fn returnStruct() -> Foo {
    let a: Foo
    a
}
fn main() {
    let x: Foo
    x.a
    returnStruct().a
    let b: Bar
    b.foo.b()
    b.foo.a
}
struct Foo {
    let a: int
    fn b() {}
}
struct Bar {
    let foo: Foo
}
"#,
        ],
        [
            "self",
            r#"
struct Foo {
    let a: int
    fn b() -> int { self.a + 1 }
}
"#,
        ],
    ];
    run_insta!("struct", tests);
}

#[test]
fn test_init_literal() {
    let tests = [
        [
            "basic",
            r#"
fn main() {
    let x: int
    let x: int8
    let x: int16
    let x: int32
    let x: int64
    let x: uint
    let x: uint8
    let x: uint16
    let x: uint32
    let x: uint64
    let x: float
    let x: double
    let x: char
    let x: bool
    let x: [bool; 3]
}
"#,
        ],
        [
            "struct",
            r#"
fn main() {
    let x: Foo
}
struct Foo {
    let a: int
    let b: bool
}
"#,
        ],
        [
            "nested_struct",
            r#"
fn main() {
    let x: Foo
}
struct Foo {
    let a: int
    let b: bool
}
struct Bar {
    let a: Foo
}
"#,
        ],
    ];
    run_insta!("init_literal", tests);
}
