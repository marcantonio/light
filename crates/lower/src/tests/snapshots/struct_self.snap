---
source: crates/lower/src/tests.rs
assertion_line: 170
expression: "(test[1], res)"
---
- "\nstruct Foo {\n    let a: int\n    fn b() -> int { self.a + 1 }\n}\n"
- Ok:
    nodes:
      - kind:
          Fn:
            proto:
              name: "_main::Foo_b~int32"
              args:
                - - self
                  - Comp: "main::Foo"
              ret_ty: Int32
              is_extern: false
              module: main
            body:
              kind:
                Block:
                  list:
                    - kind:
                        BinOp:
                          op: Add
                          lhs:
                            kind:
                              FSelector:
                                comp:
                                  kind:
                                    Ident:
                                      name: self
                                      ty:
                                        Comp: "main::Foo"
                                idx: 0
                                ty: Int32
                          rhs:
                            kind:
                              Lit:
                                value:
                                  Int32: 1
                                ty: Int32
                          ty: Int32
                  ty: Int32
    prototypes:
      - name: "_main::Foo_b~int32"
        args:
          - - self
            - Comp: "main::Foo"
        ret_ty: Int32
        is_extern: false
        module: main

